packages/idempotency/src/IdempotencyHandler.ts

```diff
@@ -313,11 +313,17 @@ export class IdempotencyHandler<Func extends AnyFunction> {
         );
       } catch (e) {
         if (e instanceof IdempotencyItemAlreadyExistsError) {
-          const idempotencyRecord: IdempotencyRecord =
-            e.existingRecord ||
-            (await this.#persistenceStore.getRecord(
+          let idempotencyRecord = e.existingRecord;
+          if (idempotencyRecord !== undefined) {
+            this.#persistenceStore.validatePayload(
```

Comment:
I do think a little bit of refactoring might be wise here, because any time changes are made to `BasePersistenceLayer.getRecord()` they will need to be duplicated in this code branch of `IdempotencyHandler.saveInProgressOrReturnExistingResult()`. I don't really have a good solution to that yet other than just making some new method on `BasePersistenceLayer` which gets called in both places and encapsulates this logic.

Reply:
Hi @kevin-secrist thank you for the detailed proposal and also for spotting the mistake.

You are correct and indeed both the previous and proposed implementations were bypassing the cache mechanism as well.

I initially considered exposing the `BasePersistenceLayer.saveToCache()` method but I ended up reverting the `BasePersistenceLayer.validatePayload()` to `private` as well as I wasn't comfortable with that in the first place.

I found your argument about duplicating implementation and future cognitive load of maintaining it in two places compelling and so I have created a new `public BasePersistenceLayer.validateExistingRecord()` method that is called both in the `IdempotencyHandler.#saveInProgressOrReturnExistingResult()` method and in the `BasePersistenceLayer.getRecord()`.

I wasn't 100% convinced with this solution either, but like you said, something different would require a more significant refactor which we can't afford to prioritize at this stage.

cc @heitorlessa - please take a look at the above and let me know if it checks out with you. This is is slightly different from what we discussed on Friday but I think it's a good option.

Reply:
happy with the solution :) That said, the naming `validateExistingRecord` may mislead a new maintainer into believing that only existing records are validated.. when in fact, it happens on both existing and `getRecord`.

Perhaps `processExistingRecord`?

Or, if your new function returns an `IdempotencyRecord`, then even `buildFromExistingRecord` would also work from a chain of responsibilities pov.

Pick what makes you more comfortable to maintain.

Reply:
I decided to apply your last suggestion, it makes a lot of sense!